File: app/src/test/java/ie/napkin/supertasks/MarkdownTest.kt
package ie.napkin.supertasks

import ie.napkin.supertasks.data.format.Markdown
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * One rule for what emphasis is, read two ways.
 *
 * A screen that can style keeps the markers and dims them; a widget or a notification cannot style
 * at all and needs them gone, because `**ship it**` with visible asterisks is worse than either
 * choice. Both read [Markdown.runs], so the interesting property is not any single output but that
 * the two can never disagree about where the emphasis was.
 */
class MarkdownTest {

    @Test
    fun `markers are removed and the words kept`() {
        assertEquals("ship it", Markdown.strip("**ship it**"))
        assertEquals("ship it", Markdown.strip("*ship it*"))
        assertEquals("ship it", Markdown.strip("`ship it`"))
        assertEquals("really ship it now", Markdown.strip("really **ship it** now"))
    }

    @Test
    fun `several runs in one line`() {
        assertEquals(
            "bold and italic and code",
            Markdown.strip("**bold** and *italic* and `code`"),
        )
    }

    @Test
    fun `bold wins over the italic that lives inside it`() {
        // The claim logic: `*italic*` must not match the inner asterisks of `**bold**`, or stripping
        // would leave a stray marker behind.
        assertEquals("bold", Markdown.strip("**bold**"))
        assertTrue(!Markdown.strip("**bold**").contains("*"))
    }

    @Test
    fun `code wins over emphasis inside it`() {
        // Inside backticks the asterisks are content, not markers, and must survive — including a
        // pair that would otherwise read as bold, which is the whole reason someone writes it in a
        // code span in the first place.
        assertEquals("a *b* c", Markdown.strip("`a *b* c`"))
        assertEquals("**not bold**", Markdown.strip("`**not bold**`"))
    }

    @Test
    fun `nested emphasis is stripped all the way down`() {
        // The outer run alone is not enough: stopping there leaves exactly the punctuation this is
        // meant to remove.
        assertEquals("both", Markdown.strip("***both***"))
        assertEquals("a b c", Markdown.strip("**a *b* c**"))
    }

    @Test
    fun `text with no emphasis is returned untouched`() {
        val plain = "buy milk tomorrow"
        assertTrue(plain === Markdown.strip(plain))
        assertEquals(emptyList<Markdown.Run>(), Markdown.runs(plain))
    }

    @Test
    fun `lone or unmatched markers are left alone`() {
        // A task called "2 * 3" is arithmetic, not emphasis, and a half-typed "**bold" is a title
        // someone is still writing.
        listOf("2 * 3", "**bold", "an * asterisk", "a ` backtick").forEach {
            assertEquals("changed: $it", it, Markdown.strip(it))
        }
    }

    @Test
    fun `stripping never leaves a marker behind`() {
        listOf(
            "**a** *b* `c`",
            "***both***",
            "**a *b* c**",
        ).forEach { line ->
            val out = Markdown.strip(line)
            // Whatever the parse decides, what reaches a widget must never show punctuation the
            // writer intended as formatting.
            assertTrue("markers survived in '$line' -> '$out'", !out.contains("**"))
        }
    }

    @Test
    fun `runs report where the words are, so a styler and a stripper agree`() {
        val text = "really **ship it** now"
        val run = Markdown.runs(text).single()
        assertEquals(Markdown.Kind.BOLD, run.kind)
        assertEquals("**ship it**", text.substring(run.outer.first, run.outer.last + 1))
        assertEquals("ship it", text.substring(run.inner.first, run.inner.last + 1))
    }
}

File: web/src/format/markdown.test.ts
import { describe, expect, it } from 'vitest'
import { plain, runs, spans } from './markdown'

/** Ported from `MarkdownTest.kt`. Same reason as the codec suite: the two clients must agree. */

const emphasis = (text: string) =>
  runs(text).map((r) => `${r.kind}:${text.slice(r.innerStart, r.innerEnd)}`)

/** Every index occupied by a run's markers. */
const markerIndices = (text: string) => {
  const out = new Set<number>()
  for (const r of runs(text)) {
    for (let i = r.outerStart; i < r.innerStart; i++) out.add(i)
    for (let i = r.innerEnd; i < r.outerEnd; i++) out.add(i)
  }
  return out
}

describe('finding emphasis', () => {
  it('finds the three kinds and the words inside them', () => {
    expect(emphasis('**ship it**')).toEqual(['bold:ship it'])
    expect(emphasis('*ship it*')).toEqual(['italic:ship it'])
    expect(emphasis('`ship it`')).toEqual(['code:ship it'])
    expect(emphasis('really **ship it** now')).toEqual(['bold:ship it'])
  })

  it('finds several runs in one line', () => {
    expect(emphasis('**bold** and *italic* and `code`'))
      .toEqual(['bold:bold', 'italic:italic', 'code:code'])
  })

  it('lets bold win over the italic that lives inside it', () => {
    // `*italic*` must not match the inner asterisks of `**bold**`, or the pair splits across two
    // runs and one asterisk of each end is left over.
    expect(emphasis('**bold**')).toEqual(['bold:bold'])
  })

  it('lets code win over emphasis inside it', () => {
    expect(emphasis('`a *b* c`')).toEqual(['code:a *b* c'])
    expect(emphasis('`**not bold**`')).toEqual(['code:**not bold**'])
  })

  it('finds nested emphasis all the way down', () => {
    expect(emphasis('***both***')).toEqual(['boldItalic:both'])
    expect(emphasis('**a *b* c**')).toEqual(['bold:a *b* c', 'italic:b'])
  })

  it('finds nothing in text with no emphasis', () => {
    expect(runs('buy milk tomorrow')).toEqual([])
  })

  it('treats lone or unmatched markers as content', () => {
    // "2 * 3" is arithmetic and "**bold" is a line someone is still typing.
    for (const line of ['2 * 3', '**bold', 'an * asterisk', 'a ` backtick']) {
      expect(runs(line), `claimed something in: ${line}`).toEqual([])
    }
  })

  it('leaves no marker of a claimed pair unclaimed', () => {
    // If the parse decides a line has emphasis, every asterisk and backtick belonging to it must
    // sit inside some run's markers. One left over is drawn at full strength beside dimmed twins.
    for (const line of ['**a** *b* `c`', '***both***', '**a *b* c**', '***a* b**']) {
      const claimed = markerIndices(line)
      const inCode = new Set<number>()
      for (const r of runs(line)) {
        if (r.kind === 'code') for (let i = r.innerStart; i < r.innerEnd; i++) inCode.add(i)
      }
      ;[...line].forEach((ch, i) => {
        if (ch === '*' || ch === '`') {
          expect(claimed.has(i) || inCode.has(i), `unclaimed ${ch} at ${i} in '${line}'`).toBe(true)
        }
      })
    }
  })

  it('reports both the markers and the words', () => {
    const text = 'really **ship it** now'
    const r = runs(text)[0]!
    expect(r.kind).toBe('bold')
    expect(text.slice(r.outerStart, r.outerEnd)).toBe('**ship it**')
    expect(text.slice(r.innerStart, r.innerEnd)).toBe('ship it')
  })
})

describe('stripping to plain', () => {
  it('agrees with the Kotlin port, case for case', () => {
    expect(plain('Buy milk')).toBe('Buy milk')
    expect(plain('*urgent*')).toBe('urgent')
    expect(plain('**urgent**')).toBe('urgent')
    expect(plain('***urgent***')).toBe('urgent')
    expect(plain('`urgent`')).toBe('urgent')
    expect(plain('Ship the **beta** today')).toBe('Ship the beta today')
    expect(plain('*red* and **green**')).toBe('red and green')
    expect(plain('2 * 3 = 6')).toBe('2 * 3 = 6')
    expect(plain('does *it')).toBe('does *it')
    expect(plain('`2 * 3`')).toBe('2 * 3')
  })

  it('unwraps only the outermost run, exactly as the renderer does', () => {
    expect(plain('**a *b* c**')).toBe('a *b* c')
  })
})

describe('spans for drawing', () => {
  it('splits a line into text and emphasis', () => {
    expect(spans('really **ship it** now')).toEqual([
      { text: 'really ' },
      { kind: 'bold', children: [{ text: 'ship it' }] },
      { text: ' now' },
    ])
  })

  it('draws only the outermost run, leaving the inner markers visible', () => {
    // Not an omission. `stripEmphasis` on Android skips nested runs too, so the phone draws this
    // bold with the asterisks around `b` still showing — and `plain` leaves them in. A renderer
    // that unwrapped one level deeper would disagree with both.
    expect(spans('**a *b* c**')).toEqual([{ kind: 'bold', children: [{ text: 'a *b* c' }] }])
  })

  it('keeps a code span’s contents literal', () => {
    expect(spans('`a *b* c`')).toEqual([{ kind: 'code', children: [{ text: 'a *b* c' }] }])
  })

  it('leaves plain text as one span', () => {
    expect(spans('buy milk')).toEqual([{ text: 'buy milk' }])
  })

  /** The invariant that keeps the two halves honest. */
  it('draws exactly the characters plain() would leave behind', () => {
    const flat = (ss: ReturnType<typeof spans>): string =>
      ss.map((s) => ('text' in s ? s.text : flat(s.children))).join('')
    for (const line of ['**a *b* c**', '`a *b* c`', 'x **y** z', '***both***', '2 * 3']) {
      expect(flat(spans(line))).toBe(plain(line))
    }
  })
})
